source/Serial/DfRobotBleSerial.cpp

```diff
@@ -0,0 +1,304 @@
+        }
+
+        return connectToDeviceAsync(_device);
+    }).then([this](Concurrency::task<void> t)
+    {
+        try
+        {
+            t.get();
+        }
+        catch (Platform::Exception ^e)
+        {
+            ConnectionFailed(L"DfRobotBleSerial::connectToDeviceAsync failed with a Platform::Exception type. (message: " + e->Message + L")");
+        }
+        catch (...)
+        {
+            ConnectionFailed(L"DfRobotBleSerial::connectToDeviceAsync failed with a non-Platform::Exception type. (name: " + _device_name + L")");
+        }
+    });
+}
+
+bool
+DfRobotBleSerial::connectionReady(
+    void
+    )
+{
+    return static_cast<bool>(InterlockedAnd(&_connection_ready, true));
+}
+
+/// \ref https://social.msdn.microsoft.com/Forums/windowsapps/en-US/961c9d61-99ad-4a1b-82dc-22b6bd81aa2e/error-c2039-close-is-not-a-member-of-windowsstoragestreamsdatawriter?forum=winappswithnativecode
+void
+DfRobotBleSerial::end(
+    void
+    )
+{
+    std::lock_guard<std::mutex> lock(_q_lock);
+    InterlockedAnd(&_connection_ready, false);
+    if (_gatt_characteristic) _gatt_characteristic->WriteClientCharacteristicConfigurationDescriptorAsync(GattClientCharacteristicConfigurationDescriptorValue::None);
+
+    // Reset with respect to dependencies
+    _rx = std::queue<byte>();
+    delete(_tx); //_tx->Close();
+    _tx = nullptr;
+    _gatt_characteristic = nullptr;
+    _gatt_service = nullptr;
+    _gatt_device = nullptr;
+    _device_collection = nullptr;
+}
+
+/// \details An Advanced Query String is constructed based upon paired bluetooth GATT devices. Then a collection is returned of all devices matching the query.
+/// \ref https://msdn.microsoft.com/en-us/library/aa965711(VS.85).aspx
+/// \warning Must be called from UI thread
+Windows::Foundation::IAsyncOperation<Windows::Devices::Enumeration::DeviceInformationCollection ^> ^
+DfRobotBleSerial::listAvailableDevicesAsync(
+    void
+    )
+{
+    // Construct AQS String for Bluetooth LE devices
+    Platform::String ^device_aqs = Windows::Devices::Bluetooth::GenericAttributeProfile::GattDeviceService::GetDeviceSelectorFromUuid(Windows::Devices::Bluetooth::GenericAttributeProfile::GattServiceUuids::GenericAccess);
+
+    // Identify all paired devices satisfying query
+    return Windows::Devices::Enumeration::DeviceInformation::FindAllAsync(device_aqs);
+}
+
+uint16_t
+DfRobotBleSerial::read(
+    void
+    )
+{
+    uint16_t c = static_cast<uint16_t>(-1);
+
+    if (available()) {
+        std::lock_guard<std::mutex> lock(_q_lock);
+        c = _rx.front();
+        _rx.pop();
+    }
+
+    return c;
+}
+
+uint32_t
+DfRobotBleSerial::write(
+    uint8_t c_
+    )
+{
+    // Check to see if connection is ready
+    if (!connectionReady()) { return 0; }
+
+    _tx->WriteByte(c_);
+    create_task(_gatt_characteristic->WriteValueAsync(_tx->DetachBuffer(), GattWriteOption::WriteWithResponse))
```

Comment:
We have discovered this is a batching error, so I'm separating the flush() operation from the write.